### ISSUE TITLE
Non-fluent API to create more generic documents

### ISSUE THREAD

**Issue by dropsonic** [QuestPDF/QuestPDF#203]
**Is your feature request related to a problem? Please describe.**
Sometimes it is required to solve more generic tasks than the fluent syntax API can address.
For example, you may have a generic in-memory object model of a report that is then rendered to HTML (web UI), Microsoft Excel (OpenXML), XML, PDF, etc.
The exact layout&content of the report is 100% configurable by the end-user.

In this case, using the fluent syntax API doesn't help.

**Describe the solution you'd like**
It would be beneficial to have full access to the internal document object model instead of the fluent syntax API.

**Comment by atliuhui** [comment]
Like WordprocessingML, QuestPDF Element objects can be serialized and deserialized.

**Comment by Poker-sang** [comment]
> Is there any operation that is currently not supported by the FluentAPI? Or is there any use case that would greatly benefit by non-fluent API?

@MarcinZiabek I have some requirements to build pdfs while parsing documents, and while the code still works now, I have to build the documents in a twisted way due to only being able to use FluentAPI.

<https://github.com/Pixeval/Pixeval/blob/3c4813bc9f93ee0a3222ca3aebf3a5c0ed2391cd/src/Pixeval/Controls/DocumentViewer/PixivNovelPdfParser.cs>

As you can see, I need to keep the last `TextDescriptor` and put it in the delegate in order to delay the execution, as it's the only way to add it to the pdf document. I may also have to use reflection if the requirements are more complex.

The library used for parsing C#, _Roslyn_, also provides FluentAPI through `SyntaxFactory`, but it also provides a way to build with new, so it's very flexible.

Taking a step back, even if the APIs don't provide `new`, you can expose the `Descriptor`s' Children/Items properties for subsequent modification as _Roslyn_ does, which also makes the APIs more flexible.

I'm very grateful to you for making QuestPDF such a great library, so I hope this suggestion will make it even better.

**Comment by MarcinZiabek** [comment]
@Poker-sang Are you sure that you need to use your complex approach? In the FluentAPI syntax, you can use control structures such as loops and conditions. For example:

```csharp
.Table(table =>
{
    table.ColumnsDefinition(columns =>
    {
        // use conditions to define dynamic columns
        if (addCustomColumnCondition)
            columns.ConstantColumn(100);  
        
        // use loops to dynamically define table's columns
        foreach (var i in Enumerable.Range(0, 4))
        {
            columns.RelativeColumn();   
        }
    });
    
    foreach (var item in Items)
    {
        if (addCustomColumnCondition)
            table.Cell().Element(Block).Text(item.Id);
        
        table.Cell().Text(item.PropertyA);
        table.Cell().Text(item.PropertyB);
        table.Cell().Text(item.PropertyC);
        table.Cell().Text(item.PropertyD);
    }
});
```

or:

```csharp
.Padding(10)
.Element(container =>
{
    if (invoice.LogoPath != null)
    {
        container.Height(50).Image(invoice.LogoPath);
    }
    else
    {
        container
            .Height(50).Width(400)
            .Background(Colors.Grey.Lighten3)
            .AlignCenter().AlignMiddle()
            .Text(invoice.CompanyName);
    }
});
```

and then you can refactor your code:

```csharp
static void AddImageLogo(IContainer container, string logoPath)
{
    container.Height(50).Image(invoice.LogoPath);
}

static void AddCustomLogo(IContainer container, string companyName)
{
    container
        .Height(50).Width(400)
        .Background(Colors.Grey.Lighten3)
        .AlignCenter().AlignMiddle()
        .Text(invoice.CompanyName);
}

.Padding(10)
.Element(container =>
{
    if (invoice.LogoPath != null)
    {
        AddImageLogo(container, invoice.LogoPath);
    }
    else
    {
        AddCustomLogo(container, invoice.CompanyName);
    }
});
```

Based on my experience, loops and condition blocks are enough to transform any generic content into dynamic document.

**Comment by girlpunk** [comment]
You can add multiple strings to a single TextDescriptor using it's Line and Span functions

**Comment by Poker-sang** [comment]
> You can add multiple strings to a single TextDescriptor using it's Line and Span functions

That's what I call a twisted approach. I have to keep the last TextDescriptor and put it in the delegate in order to delay the invokation.

**Comment by MarcinZiabek** [comment]
The suggestion provided by @girlpunk is correct.

```csharp
.Text(text =>
{
    if (showLabel)
        text.Span($"{item.Label}: ").Bold();

    text.Span($"{item.Value}");
    
    foreach (var i in Enumerable.Range(0, 4))
        text.Span(i.ToString());
});
```

**Comment by Poker-sang** [comment]
This approach requires that the TextDescriptor be placed in a lambda context, but this requirement is not always met.

Using your code as an example, I parsed a small section of text in the external, by calling `Add` I added it to the pdf document, at this time there is no problem.
```cs
void Add()
{
    xxx.Text(text =>
    {
        if (showLabel)
            text.Span($"{item.Label}: ").Bold();
    });
}
```
Then I need to go back out of the method(`Add`) and continue the Parse. after the Parse I still want to use this `TextDescriptor`, but I can't get it anymore.

In conclusion, if I want to use this approach, I can only use a queue to record the elements that need to be inserted, or use a delegate to delay the execution, which I think are both twisted.

**Comment by MarcinZiabek** [comment]
I am still trying to determine what requirement forces you to follow this workflow.

The QuestPDF-specific code should be treated as a visual layer where all data is already available. In such a case, a proper approach is first to parse your file, store all content inside a predefined set of objects, and then provide the entire state to the PDF generation logic.

Trying to combine parsing/loading logic with the generation process will not only make your life way more difficult, but also you will not gain much in terms of performance and memory usage.

**Comment by Poker-sang** [comment]
I need to generate pdf, md, html, etc. from the source file with the same code. that's why it's so decoupled.

I'm sorry to hear that. Actually my current code works, so my need for a non-fluent API is not that urgent. So I'm just raising a possibility here to make Quest-pdf more flexible and better like Roslyn.

**Comment by MarcinZiabek** [comment]
At this time, I don’t plan to introduce a non-fluent API, but I’m open to revisiting the idea in the future. That said, I’d be happy to discuss and explore ways to improve the current fluent-based development workflow and make it more flexible!

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, please feel free to open a new GitHub issue. Thanks for your feedback! ❤️
